Hevc enc sfc

feiwan1 opened the pull request.

feiwan1 committed:
lavu: add sub frame side data

Sub frame side data allows attach another AVFrame as side data into
the target AVFrame.

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 committed:
lavc: add sub frame options and flag

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 committed:
lavc/hevc_vaapi: enable sub frame support

Intel HW provide a feature that allows decoder output another scaled
frame beside original frame. And the scaled frame will attach to main
frame as sub frame side data.

The use case is mainly for video analysis. For example, scaled down
frame can be used for analysis, and the result can be applied back
to main frame.

Normally, we use scale_vaapi for scaling in vaapi transcode pipeline
if want to get a smaller resolution frame. While now sub frame can
be used instead. For some platforms, the sub frame scaling is much
more faster than scale_vaapi. For example, the decode + sub frame
cmd will improve ~50% performance than decode + scaling on my DG2
i3-11100B@3.6GHz.

decode + sub frame cmd:
ffmpeg -hwaccel vaapi -hwaccel_device /dev/dri/renderD128            \
-hwaccel_output_format vaapi -export_side_data sub_frame             \
-sub_frame_opts "width=300:height=300:format=nv12"                   \
-i 1920x1080.h265 -f null - &

decode + scaling cmd:
ffmpeg -hwaccel vaapi -hwaccel_device /dev/dri/renderD128            \
-hwaccel_output_format vaapi -i 1920x1080.h265                       \
-vf 'scale_vaapi=w=300:h=300:format=nv12' -f null - &

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 commented:
Hi @xuguangxin @xhaihao @wenbinc-Bin , could you help to review this patch? thanks.

xuguangxin commented:
@tong1wu @wenbinc-Bin could you help review this?
thanks

wenbinc-Bin commented on libavutil/sub_frame_metadata.c:
```diff
@@ -0,0 +1,70 @@
+#include "buffer.h"
+#include "common.h"
+#include "frame.h"
+#include "mem.h"
```
Is all these include needed?

wenbinc-Bin commented on libavcodec/avcodec.h:
```diff
@@ -361,6 +361,13 @@ typedef struct RcOverride{
+ * Decoding only.
+ * export sub frame through frame side data.
+ */
+#define AV_CODEC_EXPORT_DATA_SUB_FRAME (1 << 3)
```
These should be (1<<4 )?

wenbinc-Bin commented on libavcodec/avcodec.h:
```diff
@@ -2006,6 +2013,8 @@ typedef struct AVCodecContext {
      */
     int export_side_data;
 
+    AVDictionary *sub_frame_opts;
```
It is a public API. I think you should add detailed comment for it.

wenbinc-Bin commented on libavutil/hwcontext_vaapi.c:
```diff
@@ -605,6 +659,24 @@ static int vaapi_frames_init(AVHWFramesContext *hwfc)
         }
     }
 
+        desc = vaapi_format_from_pix_fmt(hwfc->sfc_format);
```
There is an indent issue.

wenbinc-Bin commented on libavcodec/vaapi_decode.c:
```diff
@@ -569,6 +587,22 @@ static int vaapi_decode_make_config(AVCodecContext *avctx,
+                else if (!strcmp(e->key, "format"))
+                    frames->sfc_format = av_get_pix_fmt(e->value);
+            }
+        }
```
If the user only set height and width will be 0, Is this configure still acceptable by libva? Or we can set default sfc_w/h equals to original w/h?

wenbinc-Bin commented on libavutil/hwcontext.h:
```diff
@@ -227,6 +228,10 @@ typedef struct AVHWFramesContext {
      */
     int width, height;
+
+    int enable_sfc;
```
Here need document as well, as it is a public API.

wenbinc-Bin commented:
I see you changed AVHWFramesContext, which may be diffcult to be accept by community. How about set a hw_frame_context in AVVAAPIFramesContext and use this frame_context to create sub_frame, so that there will be less code changing and you don't have to change AVHWFramesContext. 😁

wenbinc-Bin commented on libavcodec/vaapi_decode.h:
```diff
@@ -45,11 +45,16 @@ typedef struct VAAPIDecodePicture {
 typedef struct VAAPIDecodeContext {
     VAConfigID            va_config;
     VAContextID           va_context;
+    VAProfile             va_profile;
```
Is this variable used?

tong1wu commented on libavcodec/vaapi_decode.c:
```diff
@@ -569,6 +587,22 @@ static int vaapi_decode_make_config(AVCodecContext *avctx,
+        }
+
+        frames->enable_sfc =
+            !!((avctx->export_side_data & AV_CODEC_EXPORT_DATA_SUB_FRAME) &&
```
What is the purpose of two "!" here

feiwan1 replied:
This is used to turn none-zero value to 1. I add another condition with &&, so now don't need it anymore. thanks.

feiwan1 commented:
> I see you changed AVHWFramesContext, which may be diffcult to be accept by community. How about set a hw_frame_context in AVVAAPIFramesContext and use this frame_context to create sub_frame, so that there will be less code changing and you don't have to change AVHWFramesContext. 😁

Actually, AVHWFramesContext.hwctx is AVVAAPIFramesContext . So it shoud not add AVHWFramesContext into AVVAAPIFramesContext. If community can accept introduce sub frame concept into AVCodecContext, then they should not object related change in AVHWFramesContext, which is element of AVCodecContext.

wenbinc-Bin commented:
> > I see you changed AVHWFramesContext, which may be diffcult to be accept by community. How about set a hw_frame_context in AVVAAPIFramesContext and use this frame_context to create sub_frame, so that there will be less code changing and you don't have to change AVHWFramesContext. 😁
> 
> Actually, AVHWFramesContext.hwctx is AVVAAPIFramesContext . So it shoud not add AVHWFramesContext into AVVAAPIFramesContext. If community can accept introduce sub frame concept into AVCodecContext, then they should not object related change in AVHWFramesContext, which is element of AVCodecContext.

Sorry I didn't express correctly. I mean we can add av_buffer_ref to store it, like child_frames_ref in QSVFramesContext.

feiwan1 commented:
> > > I see you changed AVHWFramesContext, which may be diffcult to be accept by community. How about set a hw_frame_context in AVVAAPIFramesContext and use this frame_context to create sub_frame, so that there will be less code changing and you don't have to change AVHWFramesContext. 😁
> > 
> > 
> > Actually, AVHWFramesContext.hwctx is AVVAAPIFramesContext . So it shoud not add AVHWFramesContext into AVVAAPIFramesContext. If community can accept introduce sub frame concept into AVCodecContext, then they should not object related change in AVHWFramesContext, which is element of AVCodecContext.
> 
> Sorry I didn't express correctly. I mean we can add av_buffer_ref to store it, like child_frames_ref in QSVFramesContext.

Thanks Wenbin. Seems it is a feasible way. I will check if it can be easliy implemented in this patch.

feiwan1 committed:
examples: seperate vaapi_decode from hw_decode

Now vaapi_decode can be used to test vaapi decode and vaapi decode
with sub frame.

decode example:
$ vaapi_decode vaapi 1920x1080.h265 out.yuv

decode with sub frame example:
$ vaapi_decode vaapi 1920x1080.h265 out.yuv width=640:height=480:format=argb sub_out.argb

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 commented:
hi @xhaihao @xuguangxin @wenbinc-Bin @tong1wu , could you help to review this version? Thanks.
Change:
1. rebase to master.
2. maintain AVHWFramesContext for sub frame instead of just add a new frame pool for sub frame. This will avoid to change AVHWFramesContext, all changes can be done in hwcontext_vaapi.h/c.
3. add vaapi_decode example to test sfc feature.

xuguangxin commented on libavcodec/vaapi_decode.c:
```diff
@@ -518,8 +525,23 @@ static int vaapi_decode_make_config(AVCodecContext *avctx,
+                   "attributes: %d (%s).\n", vas, vaErrorStr(vas));
+            return AVERROR_EXTERNAL;
+        } else if (attr.value | VA_DEC_PROCESSING) {
+            support_sfc = 1;
```
why do we name it sfc? SFC is just the intel hw name. better follow the libva name  VA_DEC_PROCESSING

xuguangxin commented on libavcodec/vaapi_decode.c:
```diff
@@ -564,10 +586,32 @@ static int vaapi_decode_make_config(AVCodecContext *avctx,
+            avfc->sub_frame_sw_format = AV_PIX_FMT_NV12;
+            if (avctx->sub_frame_opts) {
+                AVDictionaryEntry *e = NULL;
+                while ((e = av_dict_get(avctx->sub_frame_opts, "", e, AV_DICT_IGNORE_SUFFIX))) {
```
how do we support multiple sub frames ?

feiwan1 replied:
The cmdline can be set as "-sub_frame_opts width=640:height=480:format=nv12:width=320:height=240:format=nv12....".
And here we can loop the options and see how many width/height/format can get. 
AV_DICT_IGNORE_SUFFIX this flag return the first AVDictionaryEntry and then remove it from the list.

xuguangxin commented on libavcodec/vaapi_decode.h:
```diff
@@ -45,6 +45,10 @@ typedef struct VAAPIDecodePicture {
     VABufferID           *slice_buffers;
     int                   slices_allocated;
+
+    VASurfaceID           sfc_surface;
```
better use a list for this. In the feature we may support 1:N

xuguangxin replied:
The name(sfc) may need change

feiwan1 replied:
Current code only one sub frame, we can extend it into list when we support multi sub frame.
And I will replace *sfc* in code with other string.

xuguangxin commented on libavcodec/vaapi_hevc.c:
```diff
@@ -285,6 +286,35 @@ static int vaapi_hevc_start_frame(AVCodecContext          *avctx,
             goto fail;
     }
 
+    sd = av_frame_get_side_data(h->ref->frame, AV_FRAME_DATA_SUB_FRAME);
```
can we do crop? if not the sub frame may be not a good name.
thumbnail maybe better

feiwan1 replied:
Sub frame seems better, which also can be used for audio. And thumbnail seems like a smaller picture than original picture, and it not suit if driver support scale up in the future.

xuguangxin commented on libavcodec/vaapi_decode.h:
```diff
@@ -45,6 +45,10 @@ typedef struct VAAPIDecodePicture {
     int                   slices_allocated;
+
+    VASurfaceID           sfc_surface;
+    VARectangle           src;
```
why we need this, can we put it as a local variable?

feiwan1 replied:
It's address will be set into vaCreateBuffer, and the api will copy it into a vaBuffer, so we need to keep the var before the current frame decoded.

xuguangxin commented on libavcodec/vaapi_hevc.c:
```diff
@@ -285,6 +286,35 @@ static int vaapi_hevc_start_frame(AVCodecContext          *avctx,
+        pic->pic.src.width = sps->width;
+        pic->pic.src.height = sps->height;
+
+        pic->pic.sfc.x = pic->pic.sfc.y = 0;
```
If we wang fill the entire surface, we only need put the output_region  to NULL

feiwan1 replied:
Current media-driver doesn't consider the the output_region as NULL. Will report a bug to them. And I think we can use current way to upstream ffmpeg first.

xuguangxin commented on libavutil/hwcontext_vaapi.c:
```diff
@@ -663,17 +665,92 @@ static int vaapi_frames_init(AVHWFramesContext *hwfc)
+        sub_frames_ctx->height            = avfc->sub_frame_height;
+        sub_frames_ctx->format            = AV_PIX_FMT_VAAPI;
+        sub_frames_ctx->sw_format         = avfc->sub_frame_sw_format;
+        sub_frames_ctx->initial_pool_size = hwfc->initial_pool_size;
```
It's not need preallcate

xuguangxin commented on configure:
```diff
@@ -1742,6 +1742,7 @@ EXAMPLE_LIST="
     scaling_video_example
     transcode_aac_example
     transcoding_example
+    vaapi_decode_example
```
Can we use current hw decode example to do this test?

feiwan1 replied:
If use hw_deocde do this, it will need add additional code for sub frame in hw_decode that will only available for vaapi. To avoid the community say that change will only for vaapi. So I separate it from hw_decode. I think we can send current method to community. If community doesn't mind to add sub frame code in hw_decode, I can change it back to hw_decode.

xuguangxin reviewed (commented):
EU, SFC, VDBOX are not Intel specific. Upstream may not know it. better use other word to descript it in code and commit log

xhaihao commented on libavutil/sub_frame_metadata.h:
```diff
@@ -0,0 +1,28 @@
+
+AVFrame *av_sub_frame_alloc(size_t *out_size);
+
+AVFrame *av_sub_frame_create_side_data(AVFrame *frame);
```
Please add comments for public functions and parameters

xhaihao commented on libavutil/version.h:
```diff
@@ -80,7 +80,7 @@
 #define LIBAVUTIL_VERSION_MINOR  24
-#define LIBAVUTIL_VERSION_MICRO 101
+#define LIBAVUTIL_VERSION_MICRO 102
 
```
Please bump the minor version instead of the micro version, see the requirement in https://ffmpeg.org/developer.html#Code

wenbinc-Bin commented on libavcodec/vaapi_decode.c:
```diff
@@ -564,10 +586,32 @@ static int vaapi_decode_make_config(AVCodecContext *avctx,
         frames->format = AV_PIX_FMT_VAAPI;
         frames->width = avctx->coded_width;
         frames->height = avctx->coded_height;
+        avfc->enable_sub_frame = (avctx->export_side_data & AV_CODEC_EXPORT_DATA_SUB_FRAME) && support_sfc;
```
I think "avctx->export_side_data & AV_CODEC_EXPORT_DATA_SUB_FRAME" is redundant.

xhaihao commented on libavutil/sub_frame_metadata.c:
```diff
@@ -0,0 +1,60 @@
+    av_frame_free(&frame);
+}
+
+AVFrame *av_sub_frame_alloc(size_t *out_size)
```
av_sub_frame_free() is static, please make av_sub_frame_alloc() static as well. In addition, I don't see  av_sub_frame_alloc() is used in other files.

xhaihao commented on libavutil/sub_frame_metadata.c:
```diff
@@ -0,0 +1,60 @@
+
+#include "sub_frame_metadata.h"
+
+static void av_sub_frame_free(void *opaque, uint8_t *data)
```
no prefix is required for static function, see the naming conventions in http://ffmpeg.org/developer.html#Naming-conventions

wenbinc-Bin commented on libavutil/hwcontext_vaapi.c:
```diff
@@ -663,17 +665,92 @@ static int vaapi_frames_init(AVHWFramesContext *hwfc)
+        sub_frames_ctx->sw_format         = avfc->sub_frame_sw_format;
+        sub_frames_ctx->initial_pool_size = hwfc->initial_pool_size;
+
+        ret = vaapi_hw_sub_frames_init(ctx->sub_frames_ref);
```
Why not call av_hwframe_ctx_init() and we don't need vaapi_hw_sub_frames_init() because we can reuse vaapi_hw_frames_init(). Or is there any difference I miss?

wenbinc-Bin commented on libavutil/hwcontext_vaapi.c:
```diff
@@ -683,6 +760,26 @@ static int vaapi_get_buffer(AVHWFramesContext *hwfc, AVFrame *frame)
+            return AVERROR(ENOMEM);
+
+        sub_frames_ctx = (AVHWFramesContext*)ctx->sub_frames_ref->data;
+        sub_frame->buf[0] = av_buffer_pool_get(sub_frames_ctx->pool);
```
You can call av_hwframe_get_buffer() to resue vaapi_get_buffer() and the rest code are not needed.

xhaihao commented on libavutil/sub_frame_metadata.c:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    return sub_frame;
+}
```
There is a copy of sub_frame in dst frame after calling av_frame_copy_props() to copy metadata from src to dst but this commit doesn't update all reference counts of buffers attached in AVFrame.

feiwan1 commented:
> EU, SFC, VDBOX are not Intel specific. Upstream may not know it. better use other word to descript it in code and commit log

Refined. thanks.

feiwan1 commented:
Hi @xuguangxin @xhaihao @wenbinc-Bin @tong1wu , new version pushed, could you help to review if there is any other need to change? Thanks.

xhaihao commented on libavutil/version.h:
```diff
@@ -79,7 +79,7 @@
-#define LIBAVUTIL_VERSION_MINOR  24
+#define LIBAVUTIL_VERSION_MINOR  25
 #define LIBAVUTIL_VERSION_MICRO 101
 
```
Please reset LIBAVUTIL_VERSION_MICRO to 100

xhaihao commented on libavutil/sub_frame_metadata.c:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    if (!av_frame_new_side_data_from_buf(frame, AV_FRAME_DATA_SUB_FRAME, buf)) {
+        av_buffer_unref(&buf);
```
should free sub_frame too.

feiwan1 replied:
Unref buf here will call back sub_frame_free(), so may not need to free sub_frame again. And I change freep(&sub_frame) to av_frame_free(&sub_frame) when av_buffer_create fail, which will looks better, even they are actually does same thing.

xhaihao commented on libavutil/sub_frame_metadata.h:
```diff
@@ -0,0 +1,33 @@
+ *
+ * @return The AVFrame structure to be filled by caller.
+ */
+AVFrame *av_sub_frame_create_side_data(AVFrame *frame);
```
Who'll be responsible for freeing the this frame ? I think it should be documented

xhaihao commented on libavutil/hwcontext_vaapi.c:
```diff
@@ -654,6 +655,7 @@ static int vaapi_frames_init(AVHWFramesContext *hwfc)
     }
 
     av_buffer_unref(&test_surface);
+
```
no needed.

xhaihao commented on libavcodec/avcodec.h:
```diff
@@ -2006,6 +2012,14 @@ typedef struct AVCodecContext {
+     * - encoding: unused
+     */
+    AVDictionary *sub_frame_opts;
+
```
You should add new field at the end for a public data structure.

xhaihao commented on libavutil/sub_frame_metadata.h:
```diff
@@ -0,0 +1,34 @@
+
+/**
+ * Allocate a AVFrame structure and add it to the input frame. The
+ * struct must be freed by using sub_frame_free();
```
The caller can't call sub_frame_free () because it is a static function, not a public function.

feiwan1 replied:
fixed, thanks.

feiwan1 commented:
submitted to community, close this one.
https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=6504
